File: Front-End/src/app/utils/spotify-types.ts
export interface SpotifyTracksSearchResult {
  tracks: {
    href: string;
    limit: number;
    next: string;
    offset: number;
    previous: string;
    total: number;
    items: SpotifyItem[];
  };
}

export interface SpotifyItem {
  album: SpotifyAlbum;
  artists: SpotifyArtist[];
  available_markets: string[];
  disc_number: number;
  duration_ms: number;
  explicit: boolean;
  external_ids: SpotifyExternalIds;
  external_urls: SpotifyExternalUrls;
  href: string;
  id: string;
  is_playable: boolean;
  linked_from?: any; // Replace 'any' with a more specific type if needed
  restrictions: SpotifyRestrictions;
  name: string;
  popularity: number;
  preview_url: string;
  track_number: number;
  type: string;
  uri: string;
  is_local: boolean;
}

export interface SpotifyAlbum {
  album_type: string;
  total_tracks: number;
  available_markets: string[];
  external_urls: SpotifyExternalUrls;
  href: string;
  id: string;
  images: SpotifyImage[];
  name: string;
  release_date: string;
  release_date_precision: string;
  restrictions: SpotifyRestrictions;
  type: string;
  uri: string;
  artists: SpotifyArtist[];
}

export interface SpotifyArtist {
  external_urls: SpotifyExternalUrls;
  followers?: SpotifyFollowers;
  genres?: string[];
  href: string;
  id: string;
  images?: SpotifyImage[];
  name: string;
  popularity: number;
  type: string;
  uri: string;
}

export interface SpotifyExternalUrls {
  spotify: string;
}

export interface SpotifyImage {
  url: string;
  height: number;
  width: number;
}

export interface SpotifyRestrictions {
  reason: string;
}

export interface SpotifyExternalIds {
  isrc?: string;
  ean?: string;
  upc?: string;
}

export interface SpotifyFollowers {
  href?: string;
  total: number;
}

export interface AudioFeatures {
  acousticness: number
  analysis_url: string;
  danceability: number;
  duration_ms: number;
  energy: number;
  id: string;
  instrumentalness: number;
  key: number;
  liveness: number;
  loudness: number;
  mode: number;
  speechiness: number;
  tempo: number;
  time_signature: number;
  track_href: string;
  type: string;
  uri: string;
  valence: number;
}

File: Android/app/src/main/java/com/example/whatthekey/model/SpotifyTypes.kt
package com.example.whatthekey.model

data class SpotifyTracksSearchResult(
    val tracks: Tracks
)

data class Tracks(
    val href: String,
    val limit: Int,
    val next: String,
    val offset: Int,
    val previous: String,
    val total: Int,
    val items: List<SpotifyItem>
)

data class SpotifyItem(
    val album: SpotifyAlbum,
    val artists: List<SpotifyArtist>,
    val available_markets: List<String>,
    val disc_number: Int,
    val duration_ms: Int,
    val explicit: Boolean,
    val external_ids: SpotifyExternalIds,
    val external_urls: SpotifyExternalUrls,
    val href: String,
    val id: String,
    val is_playable: Boolean,
    val linked_from: Any?, // Replace 'Any' with a more specific type if needed
    val restrictions: SpotifyRestrictions,
    val name: String,
    val popularity: Int,
    val preview_url: String,
    val track_number: Int,
    val type: String,
    val uri: String,
    val is_local: Boolean
)

data class SpotifyAlbum(
    val album_type: String,
    val total_tracks: Int,
    val available_markets: List<String>,
    val external_urls: SpotifyExternalUrls,
    val href: String,
    val id: String,
    val images: List<SpotifyImage>,
    val name: String,
    val release_date: String,
    val release_date_precision: String,
    val restrictions: SpotifyRestrictions,
    val type: String,
    val uri: String,
    val artists: List<SpotifyArtist>
)

data class SpotifyArtist(
    val external_urls: SpotifyExternalUrls,
    val followers: SpotifyFollowers?,
    val genres: List<String>?,
    val href: String,
    val id: String,
    val images: List<SpotifyImage>?,
    val name: String,
    val popularity: Int,
    val type: String,
    val uri: String
)

data class SpotifyExternalUrls(
    val spotify: String
)

data class SpotifyImage(
    val url: String,
    val height: Int,
    val width: Int
)

data class SpotifyRestrictions(
    val reason: String
)

data class SpotifyExternalIds(
    val isrc: String?,
    val ean: String?,
    val upc: String?
)

data class SpotifyFollowers(
    val href: String?,
    val total: Int
)

data class AudioFeatures(
    val acousticness: Double,
    val analysis_url: String,
    val danceability: Double,
    val duration_ms: Int,
    val energy: Double,
    val id: String,
    val instrumentalness: Double,
    val key: Int,
    val liveness: Double,
    val loudness: Double,
    val mode: Int,
    val speechiness: Double,
    val tempo: Double,
    val time_signature: Int,
    val track_href: String,
    val type: String,
    val uri: String,
    val valence: Double
)


data class RandomTrack(
    val tracks: List<Track>,
    val seeds: List<SpotifySeed>
)

data class Track(
    val album: SpotifyAlbum,
    val artists: List<SpotifyArtist>,
    val available_markets: List<String>,
    val disc_number: Int?, // Assuming this is nullable
    val duration_ms: Int?,
    val explicit: Boolean,
    val external_ids: ExternalIds,
    val external_urls: ExternalUrls,
    val href: String?,
    val id: String?,
    val is_local: Boolean,
    val name: String?,
    val popularity: Int?,
    val preview_url: String?,
    val track_number: Int?,
    val type: String?,
    val uri: String?
)
data class ExternalUrls(
    val spotify: String?
)
data class ExternalIds(
    val isrc: String?
)
data class SpotifySeed(
    val initialPoolSize: Int?,
    val afterFilteringSize: Int?,
    val afterRelinkingSize: Int?,
    val id: String?,
    val type: String?,
    val href: String?
)
